control-plane-operator/hostedclusterconfigoperator/controllers/machine/machine.go

```diff
@@ -0,0 +1,160 @@
+	var namespace string
+	if hcp.Spec.Platform.Kubevirt.Credentials != nil {
+		namespace = hcp.Spec.Platform.Kubevirt.Credentials.InfraNamespace
+	} else {
+		namespace = hcp.Namespace
+	}
+
+	// Manifests for infra/mgmt cluster passthrough service
+	cpService := hcpmanifests.IngressDefaultIngressPassthroughService(namespace)
+
+	cpService.Name = fmt.Sprintf("%s-%s",
+		hcpmanifests.IngressDefaultIngressPassthroughServiceName,
+		hcp.Spec.Platform.Kubevirt.GenerateID)
+
+	if err := r.kubevirtInfraClient.Get(ctx, client.ObjectKeyFromObject(cpService), cpService); err != nil {
+		return fmt.Errorf("failed to get default ingress passthrough Service: %w", err)
+	}
+
+	// If there is a selector endpoints should not be generated
+	if len(cpService.Spec.Selector) > 0 {
+		return nil
+	}
+
+	if len(cpService.Spec.Ports) == 0 {
+		return fmt.Errorf("missing default ingress passthrough Service %s/%s ports", cpService.Namespace, cpService.Name)
+	}
+
+	ipv4MachineAddresess := []string{}
+	ipv6MachineAddresses := []string{}
+	ports := []discoveryv1.EndpointPort{}
+
+	// If machine is back to not ready we should reset the endpointslice
+	if machine.Status.GetTypedPhase() == capiv1.MachinePhaseRunning {
+		for _, machineAddress := range machine.Status.Addresses {
+			if machineAddress.Type == capiv1.MachineInternalIP {
+				if netip.MustParseAddr(machineAddress.Address).Is4() {
+					ipv4MachineAddresess = append(ipv4MachineAddresess, machineAddress.Address)
+				} else {
+					ipv6MachineAddresses = append(ipv6MachineAddresses, machineAddress.Address)
+				}
+			}
+		}
+		for _, port := range cpService.Spec.Ports {
+			ports = append(ports, discoveryv1.EndpointPort{
+				Protocol: &port.Protocol,
+				Port:     pointer.Int32(int32(port.TargetPort.IntValue())),
+			})
+		}
+	}
+
+	if serviceHasIPFamily(cpService, corev1.IPv4Protocol) {
+		if err := r.reconcileKubevirtDefaultIngressEndpointsByIPFamily(ctx, machine, cpService, corev1.IPv4Protocol, ipv4MachineAddresess, ports); err != nil {
+			return err
+		}
+	}
+
+	if serviceHasIPFamily(cpService, corev1.IPv6Protocol) {
+		if err := r.reconcileKubevirtDefaultIngressEndpointsByIPFamily(ctx, machine, cpService, corev1.IPv6Protocol, ipv6MachineAddresses, ports); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *reconciler) reconcileKubevirtDefaultIngressEndpointsByIPFamily(ctx context.Context, machine *capiv1.Machine, cpService *corev1.Service, ipFamily corev1.IPFamily, machineAddresses []string, ports []discoveryv1.EndpointPort) error {
+	log := ctrl.LoggerFrom(ctx)
+	endpointSlice := &discoveryv1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cpService.Namespace,
+			Name:      cpService.Name + "-" + machine.Name + "-" + strings.ToLower(string(ipFamily)),
+		},
+	}
+	ownerRef := config.OwnerRefFrom(machine)
+	result, err := r.CreateOrUpdate(ctx, r.kubevirtInfraClient, endpointSlice, func() error {
```

Comment:
I just realized this owner reference on the machine won't work for us when external infra is in use. The machine and the endpointslice will be in separate OCP clusters :( 

The endpointslice is directly related to the VM though, which is in the same cluster as the endpointslice. So maybe it would make sense to use the corresponding VM as the owner.

Thankfully, looking at the Machine object, we have a good way of getting the corresponding VM quickly.

The `machine.Spec.InfrastructureRef.Name` correlates to the `KubeVirtMachine`'s name... and it just so happens that the KubeVirtMachine and VM use the same name. So the `machine.Spec.InfrastructureRef.Name` is the VM name. 

So the logic could be something like this.

```golang
result, err := r.CreateOrUpdate(ctx, r.kubevirtInfraClient, endpointSlice, func() error {
   // 1. if endpoints match machine addresses and owner ref is set, then exit early, no changes necessary and no GET on vm is required
   // 2. get VM and set owner ref on endpointslice
   // 3. set endpoints on endpointslice
}

```

Reply:
Ack, I will continue Reconciling and taking IPs from capi.Machine but using VirtualMachine as owner ref.